homeassistant/components/denonavr/config_flow.py

```diff
@@ -0,0 +1,199 @@
+        self.timeout = DEFAULT_TIMEOUT
+        self.show_all_sources = DEFAULT_SHOW_SOURCES
+        self.zone2 = DEFAULT_ZONE2
+        self.zone3 = DEFAULT_ZONE3
+        self.d_receivers = []
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initialized by the user."""
+        errors = {}
+        if user_input is not None:
+            # check if IP address is set manually
+            host = user_input.get(CONF_HOST)
+            if host:
+                self.host = host
+                return await self.async_step_settings()
+
+            # discovery using denonavr library
+            self.d_receivers = await self.hass.async_add_executor_job(denonavr.discover)
+            # More than one receiver could be discovered by that method
+            if len(self.d_receivers) == 1:
+                self.host = self.d_receivers[0]["host"]
+                return await self.async_step_settings()
+            if len(self.d_receivers) > 1:
+                # show selection form
+                return await self.async_step_select()
+
+            errors["base"] = "discovery_error"
+
+        return self.async_show_form(
+            step_id="user", data_schema=CONFIG_SCHEMA, errors=errors
+        )
+
+    async def async_step_select(self, user_input=None):
+        """Handle multiple receivers found."""
+        errors = {}
+        if user_input is not None:
+            self.host = user_input["select_host"]
+            return await self.async_step_settings()
+
+        select_scheme = vol.Schema(
+            {
+                vol.Required("select_host"): vol.In(
+                    [d_receiver["host"] for d_receiver in self.d_receivers]
+                )
+            }
+        )
+
+        return self.async_show_form(
+            step_id="select", data_schema=select_scheme, errors=errors
+        )
+
+    async def async_step_settings(self, user_input=None):
+        """Allow the user to specify settings."""
+        if user_input is not None:
+            # Get config option that have defaults
+            self.timeout = user_input[CONF_TIMEOUT]
+            self.show_all_sources = user_input[CONF_SHOW_ALL_SOURCES]
+            self.zone2 = user_input[CONF_ZONE2]
+            self.zone3 = user_input[CONF_ZONE3]
+
+            return await self.async_step_connect()
+
+        return self.async_show_form(step_id="settings", data_schema=SETTINGS_SCHEMA)
+
+    async def async_step_connect(self, user_input=None):
+        """Connect to the receiver."""
+        connect_denonavr = ConnectDenonAVR(
+            self.hass,
+            self.host,
+            self.timeout,
+            self.show_all_sources,
+            self.zone2,
+            self.zone3,
+        )
+        if not await connect_denonavr.async_connect_receiver():
+            return self.async_abort(reason="connection_error")
+        receiver = connect_denonavr.receiver
+
+        mac_address = get_mac_address(ip=self.host)
```

Comment:
If user will use hostname than IP address`get_mac_address` gives us `None`. You should use something like this https://github.com/home-assistant/core/blob/dev/homeassistant/components/huawei_lte/__init__.py#L332-L339

Reply:
I don't think hostname will work with the denonavr library anyway, we schould use IP adress I think.

Reply:
I will check if mac_address is None

Reply:
fixed

Reply:
I tested a component from your PR yesterday, I used the hostname and everything worked fine. So library works with hostnames.
Config_entry from the test:
```json
{
                "connection_class": "local_poll",
                "data": {
                    "host": "denon.bieniu.lan",
                    "receiver_id": "*AVR-X2100W-0005CD4556F4",
                    "show_all_sources": false,
                    "timeout": 2,
                    "zone2": false,
                    "zone3": false
                },
                "domain": "denonavr",
                "entry_id": "cbb47288fb934fdb9cda5a7cc16caad9",
                "options": {},
                "source": "user",
                "system_options": {
                    "disable_new_entities": false
                },
                "title": "Amplituner w salonie",
                "unique_id": "*AVR-X2100W-xxxxxxxx",
                "version": 1
            }
```

Reply:
did not know that actually worked

Reply:
I made another function to get the mac adress that now first tries to get the mac using ip, and otherwise use the hostname, so schould be fixed now